docs/encryption.md

```diff
@@ -0,0 +1,118 @@
+# containerd image encryption
+
+The containerd encryption feature allows the encryption and decryption of a container image.
+It is based on a [proposal for an OCI specification](https://github.com/opencontainers/image-spec/issues/747) to define the structure of an encrypted image.
+The encryption is specified on the layer, allowing the encryption of specific layers within an image. For example, given an image with an ubuntu, node and some custom code on top, only the custom code (top-most layer) can be encrypted to still benefit from layer deduplication of insensitive data. The key sharing is done via wrapped keys in the metadata.
+Therefore, an encrypted image data intended for multiple recipients can also be deduplicated.
+All this is done via the addition of additional layer mediatype `+enc`.
+More details can be viewed in the [design doc](https://docs.google.com/document/d/146Eaj7_r1B0Q_2KylVHbXhxcuogsnlSbqjwGTORB8iw).
+
+The two main usage points are in the creation of an encrypted image, and the decryption of the image upon usage.
+As most of the integration points would be transparent or consumed by other containerd runtime components or other utilities like buildkit, we have created several `ctr` commands to better illustrate the usage of the encryption feature.
+
+# Example End User Usage
+
+We have added 3 commands in the [`ctr`](https://github.com/containerd/containerd/tree/master/cmd/ctr) client under the image module. They are `ctr image encrypt/decrypt/layerinfo`.
+
+## Encrypt
+
+The following command performs an encryption of the image `docker.io/library/alpine:latest` to an encrypted image with the tag `docker.io/library/alpine:enc`.
+The encryption is done for two recipients with the public key of `/tmp/tmp.AGrSDkaSad/mypubkey.pem` (jwe) and `/tmp/tmp.AGrSDkaSad/clientcert.pem` (pkcs7).
+The option `--layer -1` specifies the layer filter for encryption, -1 indicating the top-most layer should be encrypted.
+
+```
+$ ctr images encrypt \
+    --recipient /tmp/tmp.AGrSDkaSad/mypubkey.pem \
+    --recipient /tmp/tmp.AGrSDkaSad/clientcert.pem \
+    --layer -1 \
+    docker.io/library/alpine:latest docker.io/library/alpine:enc
+
+Encrypting docker.io/library/alpine:latest to docker.io/library/alpine:enc
+```
+
+## Layerinfo
+
+The layerinfo command provides information about the encryption status of an image. In the following command, we use it to inspect the encryption metadata.
+
+```
+$ ctr images layerinfo docker.io/library/alpine:enc
+   #                                                                    DIGEST         PLATFORM      SIZE   ENCRYPTION   RECIPIENTS
+   0   sha256:3427d6934e7749d556be6881a17265c9817abc6447df80a09c8eecc465c5bfb3      linux/amd64   2206947
+   0   sha256:d9a094b6b49fc760501d44ae96f19284e86db0a51b979756ca8a0df4a2746c79     linux/arm/v6   2146469
+   1   sha256:ef87d8b3048d8f1f7af7605328f63aab078a1433116dc15738989551184d7a87     linux/arm/v6       191   jwe,pkcs7    [jwe], [pkcs7]
```

Comment:
Are the recipients users or are they the same as the algo?

Reply:
For some encryption technologies we can display the actual recipients, such as for example OpenPGP, for others we cannot but can only show how many there are.

Reply:
Actually, even displaying the exact number of recipients may not always be possible. 
JWE and PKCS7 work with key (files) rather than a keyring (like OpenPGP) and don't associate these keys with email addresses like OpenPGP does via its keyring.